core/common/src/main/java/alluxio/security/authentication/AuthenticatedUserInjector.java

```diff
@@ -44,39 +48,54 @@ public AuthenticatedUserInjector(AuthenticationServer authenticationServer) {
   @Override
   public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
       Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    /**
+     * For streaming calls, below will make sure authenticated user is injected prior to creating
+     * the stream.
+     */
+    if (!authenticateCall(call, headers)) {
+      return next.startCall(call, headers);
+    }
+
+    /**
+     * For non-streaming calls to server, below listener will be invoked in the same thread that is
+     * serving the call.
+     */
     return new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(
         next.startCall(call, headers)) {
-      /**
-       * onHalfClose is called on the same thread that calls the service handler.
-       */
       @Override
       public void onHalfClose() {
-        // Try to fetch channel Id from the metadata.
-        UUID channelId = headers.get(ChannelIdInjector.S_CLIENT_ID_KEY);
-        boolean callClosed = false;
-        if (channelId != null) {
-          try {
-            // Fetch authenticated username for this channel and set it.
-            String userName = mAuthenticationServer.getUserNameForChannel(channelId);
-            if (userName != null) {
-              AuthenticatedClientUser.set(userName);
-            } else {
-              AuthenticatedClientUser.remove();
-            }
-          } catch (UnauthenticatedException e) {
-            call.close(Status.UNAUTHENTICATED, headers);
-            callClosed = true;
-          }
-        } else {
-          call.close(Status.UNAUTHENTICATED, headers);
-          callClosed = true;
-        }
-        // Move the call up to the handler chain,
-        // if we have not closed the call already due to authentication error.
-        if (!callClosed) {
+        if (authenticateCall(call, headers)) {
           super.onHalfClose();
         }
       }
     };
   }
+
+  private <ReqT, RespT> boolean authenticateCall(ServerCall<ReqT, RespT> call, Metadata headers) {
+    // Try to fetch channel Id from the metadata.
+    UUID channelId = headers.get(ChannelIdInjector.S_CLIENT_ID_KEY);
+    boolean callClosed = false;
```

Comment:
Could keep track of `callAuthenticated` instead, it makes more sense with the method name and we don't need to negate it?